### PR TITLE
[WIP] [Android] Add option to manually enter Account Manager URL

### LINKED ISSUE #1582: Android - Allow adding other account managers
The Windows client allows to enter an account manager URL, so it's very flexible.
On Android (V 7.4.53) currently only 3 account managers are shown and can be used:

BAM    @   http://bam.boincstats.com/
http://gridrepublic.org/
http://www.extremadurathome.org/extremadurathome/

Would it be possible to allow adding an account manager by URL ?
(similar to the existing 'Add project by URL')

Or to include the Gridcoin.co Pool account manager @  http://pool.gridcoin.co   in the list of account managers ?

If the 2nd option is implemented please note

> How do I fix "[error] Inconsistent signing key from account manager"?
> Please make sure to use "**http://pool.gridcoin.co**" as the URL and NOT "http://pool.gridcoin.co/", this is a problem with how BOINC treats URLs.

from   http://pool.gridcoin.co/faq/#how_do_i_fix_inconsistent_signing_key_from_account_manager

### PR DESCRIPTION
**Pull request by adamradocz**
Fixes #1582
Closes #2208

**Description of the Change**
Added the option to use a custom account manager.

Currently, when an account manager is selected from the list, its URL is copied to the textbox and the textbox's value used as the URL.

**Alternate Designs**
I understand that selecting an account manager from a list is more convenient, than typing it, but IMO, the current solution looks a little bit weird, because of the duplicated URL.

What do you guys think? Should the current solution stay or should we remove the list?

**Release Notes**
[Android] Allow entering custom Account Manager URL.

screenshot_20181211-135430_boinc

### CODE REVIEW

**Comment by brevilo** [comment]
Wow, @adamradocz you're really on fire right now. Do we have a "contributor of the month" badge somewhere? If not, we probably should get one and stick it to you 😁

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/attach/AcctMgrFragment.java [hunk 1]
```diff
@@ -70,16 +72,29 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
         doBindService();
         View v = inflater.inflate(R.layout.attach_project_acctmgr_dialog, container, false);
         
-        urlSpinner = (Spinner) v.findViewById(R.id.url_spinner);
+        urlSpinner = v.findViewById(R.id.url_spinner);
```
[the remark this answers is not in the document]
Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/attach/AcctMgrFragment.java]: Thanks, next time I'll.

**Comment by adamradocz** [comment]
> From the PR's description I can't really tell what is old and what is new. From looking at the code I'd say you only added the URL input field to able to either override/adjust a selection or just entry a URL yourself.
> 
> I think your PR is indeed useful as it follows what's possible with the BOINC desktop manager. However, building on that analogy I'd consequently change the drop down list in such a way that the list entries would show as service names instead of URLs. IOW, the list entry in your example should read "BAM! / BOINCstats" instead of "https://bam.boincstats.com". Of course you'd still populate the URL field with actual URL upon selection. Would that be possible?

At the moment it's not possible, because the current implementation uses a separate config, which contains only the URLs, but not the names. It could be possible if the change mentioned here #2208 would be done.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/attach/AcctMgrFragment.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (brevilo) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/attach/AcctMgrFragment.java]: Sure, that's how I meant it 👍

**Comment by brevilo** [comment]
Ok, I see. I agree that #2208 should take care of that prerequisite. That would be a nice solution avoiding redundant list management **and** allowing to separate label from URL in this PR. So let's focus on #2208 first.

**Comment by adamradocz** [comment]
I managed to make the change to use the `all_projects_list.xml` instead the hard-coded values.
The only issue is, I really don't have clue how to implement the drop down list in such a way that the list entries would show as service names instead of URLs.

**Comment by adamradocz** [comment]
Changed the image in the first post, now it shows the current state.

**Comment by brevilo** [comment]
Nice! Glad you found a solution. I'm going to have a look at the changes on Thursday. In the meantime two questions:

1. How about the dialog title? Will it always be cut off like that? Any chance of having it wrap around? As it is no one can read it on, say, a smaller phone like your's.
1. How does your implementation affect #2208?

Thanks!

**Comment by adamradocz** [comment]
@brevilo 

1. This PR uses the default implementation, it nothing to do with the window size. To change that we have to rewrite that part of the UI. BTW I have a 6" device, so not that small :)

1. This PR closes the #2208, that use the hard-coded option. Maybe the account manager proposed there can be added the the `all_projects_list.xml`, but that should be a new PR.

**Comment by RichardHaselgrove** [comment]
In #2208 it was asserted that additions to (and by implication, subtractions from) `all_projects_list.xml` should be a (recorded) decision of the PMC, not the individual decision of one person who might have a foot in both camps.

**Comment by adamradocz** [comment]
It turned out it only works if at least one project has been added, but on fresh install it crashes.
Currently I don't know why.

**Comment by brevilo** [comment]
Ok, I'll wait with the review then. Is anyone able to able to help Adam with this final issue?

**Comment by adamradocz** [comment]
The problem is, that the `monitor` object is `null` at a fresh start, despite the fact that the `doBindService` should start it.

**Comment by brevilo** [comment]
@AenBleidd are you able to help here? I can look into this in January again...

**Comment by adamradocz** [comment]
From now on, I won't make any new PR, because it would interfere with this PR too much.

**Comment by adamradocz** [comment]
I've put countless hours resolving this issue and I'm starting to think, that the original developer couldn't manage to use the `Monitor` class, so he went an easier way.

**Comment by brevilo** [comment]
Hi Adam, I'm not an expert on these specific Android concepts. What's the actual problem right now? Looking at the history of this PR it seemed almost ready to merge until you started with the `BOINCActivity` or `bindService` stuff (d03cc5e). What's the problem you're trying to solve using those? Are these optimizations to do things "the right way" or is there a real showstopper? Cheers.

**Comment by adamradocz** [comment]
> Hi Adam, I'm not an expert on these specific Android concepts. What's the actual problem right now? Looking at the history of this PR it seemed almost ready to merge until you started with the `BOINCActivity` or `bindService` stuff ([d03cc5e](https://github.com/BOINC/boinc/commit/d03cc5e19da011a3d216082550972de5daac4274)). What's the problem you're trying to solve using those? Are these optimizations to do things "the right way" or is there a real showstopper? Cheers.

The `BOINCActivity` only initialized, when at least one project is added. Before that is unusable.
The `Monitor` service can't be initialized (at least I don't know how to do it) from `AcctMgrFragment`.

BTW, the `BOINCActivity` uses the `Monitor` service too.

The `Monitor` service is the way to get the account manager information from the `all_projects_ list.xml`

**Comment by adamradocz** [comment]
I propose to merge the part only, when we used the old version with the custom xml file, to move this issue forward. It will be half solution, but at least one half is done. :)

**Comment by brevilo** [comment]
You first reply doesn't tell me anything ;-) From the second I figure that `BOINCActivity` and `Monitor` are needed/required to fetch the list of projects from somewhere else than the `all_projects_list.xml` that you [confirmed](https://github.com/BOINC/boinc/pull/2867#issuecomment-446148933) to be possible earlier, right? If so, what is this other place and why is the XML file not working anymore, or, if it is, what's wrong about using it? Still confused...

**Comment by AenBleidd** [comment]
I could probably finish this PR. #3109 is the fix for the first issue, that list of Account Managers couldn't be filled because of uninitialized Monitor instance. I'm working on the next one, impossibility to connect to selected Account Manager if no projects we connected by the user first. It requires some time to understand all over-complicated (at least it looks this way for me now) logic to find the root of issue and ways to solve it.
Definitely I want to fix these two issues (and other if one will be found during the testing) and collect all of them to this PR and merge them together.
As soon as @adamradocz has no time in the near future to work on this PR too and probably review it also, I'd like you, @brevilo, to help me with review and test.
Thanks a lot in advance

**Comment by brevilo** [comment]
I'm sure how I can help other than doing some loose meta-review as I have no idea about Android's overall development model.

**Comment by adamradocz** [comment]
Before the merger, the coding style must be fixed.

**Comment by adamradocz** [comment]
@AenBleidd I fixed the conflicts, now you can merger your solution with this PR.

**Comment by brevilo** [comment]
@AenBleidd could you please sign off on Adam's merge of your PR #3109? Thanks.

**Comment by AenBleidd** [comment]
There is one more commit made by @adamradocz so I can't just make amend to sign-off merge commit.
Will review of this PR be enough?

**Comment by brevilo** [comment]
This took a while but it was worth it. Thanks to everyone involved!

**Pull request merged**